### PR TITLE
CBRAIN implementation for user ssh key data provider including ssh key download

### PR DESCRIPTION
**Pull request by MontrealSergiy**
also reformated the ssh key css class

### CODE REVIEW

**Comment by MontrealSergiy** [comment]
unittest fail

File: BrainPortal/app/controllers/data_providers_controller.rb [hunk 1]
```diff
@@ -84,13 +84,16 @@ def new #:nodoc:
                                 )
 
     @typelist = get_type_list
+    render template: 'data_providers/normal_new' unless current_user.has_role?(:admin_user) # normal user only allowed create UserkeyFlatDirSshDataProvider
   end
 
   def create #:nodoc:
-    @provider = DataProvider.sti_new(data_provider_params)
+    @provider = DataProvider.sti_new(data_provider_params) if current_user.has_role?(:admin_user)
```
Comment (prioux) [inline BrainPortal/app/controllers/data_providers_controller.rb:91]: You modified the create action to add support for normal users creating their own DP. The create action used to be restricted to admins, so because of that it was allowed to do a lot of things without checking for security issues. Now that normal users can invoke it, they can trigger these security issues. E.g. they can create a DP for another user or for a project that they don't have access!

Instead of trying to retrofit this original create(), set up a separate action for creating the user DPs, a bit like it was done on the neurohub side. That will also provide you with examples of what a user is allowed to set or not in the object.

Comment (prioux) [reply BrainPortal/app/controllers/data_providers_controller.rb:91]: The new action could be called "create_userdp" or something.

File: BrainPortal/app/controllers/data_providers_controller.rb [hunk 2]
```diff
@@ -819,6 +823,153 @@ def repair
     end
   end
 
+  # This action checks that the remote side of the DataProvider is
+  # accessible using SSH.
+  def check
```
Comment (prioux) [inline BrainPortal/app/controllers/data_providers_controller.rb:828]: Most of this code should be implemented in the DP model instead, and just to top level code left in the controller side.

File: BrainPortal/app/controllers/data_providers_controller.rb [hunk 3]
```diff
@@ -842,7 +993,8 @@ def data_provider_params #:nodoc:
       # they don't control.
       params.require_as_params(:data_provider).permit(
         :name, :description, :group_id, :time_zone,
-        :alternate_host,
+        :alternate_host, :remote_user, :remote_host, :alternate_host, :remote_dir,
```
Comment (prioux) [inline BrainPortal/app/controllers/data_providers_controller.rb:996]: This introduces serious security issues that could allow users total control of all CBRAIN. Don't modify security code unless you understand the implications!

**Review by prioux — CHANGES_REQUESTED** [review]
Serious security issues. Redesign completely:

* new distinct controller action for create
* distinct views for the DP form
* no need to implement the autoregister action, since we already have a registration page in CBRAIN
* do not let normal users see or modify anything they didn't have access to before!

**Comment by prioux** [comment]
No auto registration. I want users to know what files they connect in CBRAIN.

File: BrainPortal/app/controllers/data_providers_controller.rb [hunk 4]
```diff
@@ -110,6 +115,39 @@ def create #:nodoc:
...
+  # allow user create his own data providers.  At the moment only ssh with UserkeyFlatDirSshDataProvider and few field only
+  def create_by_normal_user
+
+    normal_params = params.require_as_params(:data_provider)
+                        .permit(:name, :description, :group_id,
+                                :remote_user, :remote_host,
+                                :remote_port, :remote_dir
+                                )
+    group_id = normal_params[:date_provider][:group_id]
+    Group.where(id: current_user.assignable_group_ids).find(group_id) # ensure assignable, not sure need check visibility etc more
```
Comment (prioux) [inline BrainPortal/app/controllers/data_providers_controller.rb:128]: Why are you querying the DB twice? Why not instead:

```ruby
current_user.assignable_groups.find(group_id)
```

It's not for nothing that we added all these utility methods to the `User` class for fetching groups!

File: BrainPortal/app/views/data_providers/_show_user_key.html.erb [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+      </td></tr>
+      <tr><th class="supertitle"> Command line </th></tr>
+      <tr><td>
+        <pre class="ssh_key"> cat mykey.pub | ssh REMOTE_USERNAME@REMOTE_HOSTNAME "mkdir -p ~/.ssh && chmod 700 ~/.ssh && cat >>  ~/.ssh/authorized_keys</pre>
+      </td></tr>
+      <tr><td>
+        If download did not work paste the ssh key bellow in a text file and rerun the command line
+      </tr></td>
+      <tr><th class="supertitle">Public SSH Key</th></tr>
+      <tr><td><pre class="ssh_key"><%= pretty_ssh_key @data_provider.user.ssh_key(create_it: true).public_key rescue "Error fetching public key" %></pre></td></tr>
```
Comment (prioux) [inline BrainPortal/app/views/data_providers/_show_user_key.html.erb:44]: This always crashes and never show the key, because `@data_provider` is not defined in a partial.

Again: `Did you even try your own code?!?`

File: BrainPortal/app/views/data_providers/new.html.erb [hunk 6]
```diff
@@ -223,7 +223,7 @@
 <div class="display_cell" style="padding-left: 1em;">
   <table>
     <tr><th colspan="2" class="supertitle">Public SSH Key for this CBRAIN Portal</th></tr>
-    <tr><td>This key should be installed on the host machines of SSH or Smart Data Providers to allow remote access.</td></tr>
+    <tr><td>This key should be installed on the host machines of SSH or Smart Data Providers (save User-key based) to allow remote access.</td></tr>
```
Comment (prioux) [inline BrainPortal/app/views/data_providers/new.html.erb:226]: Undo changes here. This `new` form is only for standard DPs created by the admin.

File: BrainPortal/app/views/data_providers/new_normal.html.erb [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+<h2>Add New SSH based Data Provider</h2>
+
+<%= error_messages_for(@provider) %>
+
+<%= form_for @provider, :as => :data_provider, :url => {:action => "create_for_normal_user"}, :datatype => "script" do |f| -%>
+
+  <div class="generalbox">
+
+    <span title="The name of the data provider.">
```
Comment (prioux) [inline BrainPortal/app/views/data_providers/new_normal.html.erb:35]: Since this form is meant for normal users instead of sysadmins, replace all the "title" in the spans with field_explanations.

Also, the original HTML was wrong, and all these SPANs are supposed to be DIVs.

File: BrainPortal/app/views/data_providers/new_normal.html.erb [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+        <p>
+          <%= f.label :remote_port, "Remote Port Number" %><br/>
+          <%= f.text_field :remote_port, :size => 6 %>
+        </p>
+      </div>
+    </fieldset>
+
+    <p>
+
+    <fieldset class="groupentry">
```
Comment (prioux) [inline BrainPortal/app/views/data_providers/new_normal.html.erb:125]: Remove completely the advanced params section

File: BrainPortal/app/views/data_providers/show.html.erb [hunk 9]
```diff
@@ -42,9 +42,14 @@
                   :generate         => "ok"
                 ), :class => "button" %>
 
-  <% if check_role(:admin_user) || @provider.user_id == current_user.id %>
-    <%= link_to 'Delete', data_provider_path(@provider), { :data => { :confirm => "Are you sure you want to delete '#{@provider.name}' ?" }, :method => :delete, :class => "button" } %>
-  <% end %>
+
+    <% if @provider.type == "UserkeyFlatDirSshDataProvider" %>
```
Comment (prioux) [inline BrainPortal/app/views/data_providers/show.html.erb:46]: Don't do equality checks on object types, use `.is_a?` instead.

Comment (prioux) [reply BrainPortal/app/views/data_providers/show.html.erb:46]: Also, maybe better is to verify the capability of the data provider. If it has the check() method.

File: BrainPortal/app/views/data_providers/show.html.erb [hunk 10]
```diff
@@ -259,22 +264,20 @@
...
-<P>
-
-<div>
-  <table>
-    <tr><th colspan="2" class="supertitle">Public SSH Key for this CBRAIN Portal</th></tr>
-    <tr><td>This key should be installed on this Data Provider's host machine to allow remote access.</td></tr>
-    <tr><td><PRE class="ssh_key"><%= pretty_ssh_key(RemoteResource.current_resource.get_ssh_public_key || 'Unknown! Talk to sysadmin!') %></PRE></td></tr>
-  </table>
-</div>
+<% if @provider.has_owner_access?(current_user) %>
```
Comment (prioux) [inline BrainPortal/app/views/data_providers/show.html.erb:267]: Why did you remove the `if` check of the original code? It was important! Now your code will show a system SSH key even for DPs where it is irrelevant.

File: BrainPortal/app/views/data_providers/show_custom.html.erb [hunk 11]
```diff
@@ -0,0 +1,245 @@
+
```
Comment (prioux) [inline BrainPortal/app/views/data_providers/show_custom.html.erb:1]: Delete this file completely, it's unnecessary. Like I said, the codebase's original 'show' page fully supported UserKey DPs.

File: BrainPortal/config/routes.rb [hunk 12]
```diff
@@ -115,10 +115,12 @@
       get  'report'
       post 'report'
       post 'repair'
+      get  'check'
```
Comment (prioux) [inline BrainPortal/config/routes.rb:118]: check() is not a stateless action, make it a POST.

File: BrainPortal/config/routes.rb [hunk 13]
```diff
@@ -115,10 +115,12 @@
...
     end
     collection do
       get  'dp_access'
       get  'dp_transfers'
+      post 'create_for_normal_user'
```
Comment (prioux) [inline BrainPortal/config/routes.rb:123]: `create_user_dp`

File: BrainPortal/lib/user_key_test_connection_error.rb [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+# This class provides an exception class for
+# representing a user key connection error.
+# At the moment it does not guaranty though that connection is good.
+# Currently # used on CbrainTask objects.
+class UserKeyTestConnectionError < CbrainException
```
Comment (prioux) [inline BrainPortal/lib/user_key_test_connection_error.rb:27]: 1. consider renaming (see previous comment)
2. even though you extracted the code form the NH codebase, you left the original class definition in the NH codebase too
3. Adjust the comment for the class
4. adjust the Copyright date range in the header (also do this with all other files you added)

**Review by prioux — CHANGES_REQUESTED** [review]
Significant changes.

File: BrainPortal/app/controllers/data_providers_controller.rb [hunk 15]
```diff
@@ -110,6 +115,39 @@ def create #:nodoc:
     end
   end
 
+  # dp creation by normal user, currently only UserkeyFlatDirSshDataProvider
+  # allow user create his own data providers.  At the moment only ssh with UserkeyFlatDirSshDataProvider and few field only
+  def create_by_normal_user
```
[the remark this answers is not in the document]
Comment (MontrealSergiy) [reply BrainPortal/app/controllers/data_providers_controller.rb:120]: well ok though by_normal_user is less confusing. (Admin can create a pd for user personal usage)

File: BrainPortal/app/views/data_providers/show_custom.html.erb (hunk 11 above)
Comment (MontrealSergiy) [reply BrainPortal/app/views/data_providers/show_custom.html.erb:1]: did not notice. Well, obviously be adjusted with extra embedded code, and logic in templates is harder to maintain due to mix of markup and embedded code

File: BrainPortal/app/controllers/data_providers_controller.rb [hunk 16]
```diff
@@ -125,6 +163,9 @@ def update #:nodoc:
        return
     end
 
+    Group.where(id: current_user.assignable_group_ids).find(group_id) if ! current_user.has_role?(:admin_user)
```
[the remark this answers is not in the document]
Comment (MontrealSergiy) [reply BrainPortal/app/controllers/data_providers_controller.rb:166]: that is to prevent malicious/abusive users from assigning groups they should not have access to

[the remark this answers is not in the document]
Comment (prioux) [reply BrainPortal/app/controllers/data_providers_controller.rb:166]: It doesn't check anything: the line ALWAYS crashes because `group_id` is an undefined variable Sergiy. You did not even try the code.

[the remark this answers is not in the document]
Comment (MontrealSergiy) [reply BrainPortal/app/controllers/data_providers_controller.rb:166]: I got that

[the remark this answers is not in the document]
Comment (MontrealSergiy) [reply BrainPortal/app/controllers/data_providers_controller.rb:166]: what I just wanted to state purpose of the line

File: BrainPortal/app/controllers/data_providers_controller.rb [hunk 17]
```diff
@@ -840,9 +950,14 @@ def data_provider_params #:nodoc:
       # Normal users are not allowed to change
       # some parameters that would allow them to access things
       # they don't control.
+      # remote host data are only editable on user created private storage
```
[the remark this answers is not in the document]
Comment (MontrealSergiy) [reply BrainPortal/app/controllers/data_providers_controller.rb:953]: it was used for editing dp. If user initially created a typo, changed username, port etc. Well maybe he can live without it, just trying to transfer neurohub functionalities

File: BrainPortal/app/controllers/data_providers_controller.rb [hunk 18]
```diff
@@ -110,6 +115,39 @@ def create #:nodoc:
...
+
+    normal_params = params.require_as_params(:data_provider)
+                        .permit(:name, :description, :group_id,
+                                :remote_user, :remote_host,
+                                :remote_port, :remote_dir
+                                )
+    group_id = normal_params[:date_provider][:group_id]
+    Group.where(id: current_user.assignable_group_ids).find(group_id) # ensure assignable, not sure need check visibility etc more
+    @provider = UserkeyFlatDirSshDataProvider.new(normal_params)
+    @provider.update_attributes(
```
[the remark this answers is not in the document]
Comment (MontrealSergiy) [reply BrainPortal/app/controllers/data_providers_controller.rb:130]: ok

File: BrainPortal/app/controllers/data_providers_controller.rb [hunk 19]
```diff
@@ -65,7 +65,10 @@ def show  #:nodoc:
     cb_notice "Provider not accessible by current user." unless @provider.can_be_accessed_by?(current_user)
 
     respond_to do |format|
-      format.html # show.html.erb
+      format.html {
```
[the remark this answers is not in the document]
Comment (MontrealSergiy) [reply BrainPortal/app/controllers/data_providers_controller.rb:68]: I can do it, but NeuroHub had less irrelevant fields, not to mention all edit would stick

File: BrainPortal/app/controllers/data_providers_controller.rb [hunk 20]
```diff
@@ -84,10 +87,12 @@ def new #:nodoc:
                                 )
 
     @typelist = get_type_list
+    @groups   = current_user.assignable_groups | current_user.listable_groups #
```
[the remark this answers is not in the document]
Comment (MontrealSergiy) [reply BrainPortal/app/controllers/data_providers_controller.rb:90]: not sure why make yet another attribute, is not @data_provider clear enough

File: BrainPortal/config/routes.rb (hunk 13 above)
Comment (MontrealSergiy) [reply BrainPortal/config/routes.rb]: ok

File: BrainPortal/app/views/data_providers/show.html.erb (hunk 9 above)
Comment (MontrealSergiy) [reply BrainPortal/app/views/data_providers/show.html.erb:46]: good idea, but now we should ow not use check for any other purposes and maintain aryty

File: BrainPortal/app/views/data_providers/new_normal.html.erb (hunk 7 above)
Comment (MontrealSergiy) [reply BrainPortal/app/views/data_providers/new_normal.html.erb]: ok, maybe `show table` then

**Comment by MontrealSergiy** [comment]
all comments are covered

**Comment by MontrealSergiy** [comment]
the patch is merged into the pr

**Comment by MontrealSergiy** [comment]
At this point a condition in get_type_list method might be dropped without affecting much
```
--- a/BrainPortal/app/controllers/data_providers_controller.rb
+++ b/BrainPortal/app/controllers/data_providers_controller.rb
@@ -938,11 +938,8 @@ class DataProvidersController < ApplicationController
 
   def get_type_list #:nodoc:
     data_provider_list = [ "FlatDirSshDataProvider" ]
-    if check_role(:site_manager) || check_role(:admin_user)
       data_provider_list = DataProvider.descendants.map(&:name)
-      data_provider_list.delete(UserkeyFlatDirSshDataProvider.name) # this type is for regular users
-                                                                    # not for admins
-    end
+      data_provider_list.delete(UserkeyFlatDirSshDataProvider.name)  # this type only to be used with personal dp template
     grouped_options = data_provider_list.to_a.hashed_partitions { |name| name.constantize.pretty_category_name }
     grouped_options.delete(nil) # data providers that can not be on this list return a category name of nil, so we remove them
     grouped_options.keys.sort.map { |type| [ type, grouped_options[type].sort ] }

```
